src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -3350,57 +3378,86 @@ IsPartitionColumnRecursive(Expr *columnExpression, Query *query)
 		{
 			candidateColumn = (Var *) fieldExpression;
 		}
-		else
```

Comment:
I don't know a way to get to this state. There is no regression test hitting this one. 
If somehow  we get in this case  relationId would be InvalidOid, and column would be NULL. Callers would complain about not finding a partitionColumn instead of this error.